### ISSUE TITLE
UTF-8 chars U+2028 and U+2029 are not escaped - lead to JSON parse error in JavaScript

### ISSUE THREAD

**Issue by shoteff** [FasterXML/jackson-databind#1485]
Hi,
I recently discovered a problem with the escaping in the JSON serialized value. I'm using Jackson 2.4.1, which comes by default with WildFly 8.2.1. By default only the control symbols with ASCII code below 0xF0 are escaped, but UTF-8 chars U+2028 and U+2029 are causing problems by breaking the string literal in Javascript. According to the ECMAScript specification (http://www.ecma-international.org/ecma-262/7.0/index.html#sec-literals-string-literals) those symbols are interpreted as line terminators, thus effectively breaking the string literal value (see NOTE 2). AFAIK those symbols are allowed by the general JSON specification, only causing problems in JavaScript. I managed to resolve the problem by creating a custom character escape implementation and configuring the ObjectMapper to use it, but probably it is a good idea to provide more straightforward way to set up ECMAScript compatibility (new SerializationFeature option ?) of the JSON serialized result (or even by default)

**Comment by cowtowncoder** [comment]
Yes, I see the challenge (and I thought this had been reported earlier, but couldnt see another issue).
While it should be quite easy to enable escaping with `CharacterEscapes`, there is no simple out-of-the-box way to enable this.

So... need to think of what to do here.

**Comment by Ocramius** [comment]
Similar issue from my end, but when sending data like `{"contains": ["Ümläuts"]}` to a PHP backend.

I looked at `CharacterEscapes`, but can't fully figure out how it's supposed to work: should I subclass `JsonpCharacterEscapes`?

**Comment by Ocramius** [comment]
Forgive me for the hand-holding, not familiar with the ecosystem around Java and this library.

What I attempted is to bypass the entire problem by escaping everything outside ASCII space:

```java
public class EscapeAllNonAsciiCharacters extends CharacterEscapes {
    private final CharacterEscapes next;

    public EscapeAllNonAsciiCharacters(CharacterEscapes next) {
        this.next = next;
    }

    @Override
    public int[] getEscapeCodesForAscii() {
        return next.getEscapeCodesForAscii();
    }

    @Override
    public SerializableString getEscapeSequence(int i) {
        SerializableString escaped = next.getEscapeSequence(i);

        if (escaped != null) {
            return escaped;
        }

        if (i < 256) { // I've yet to unit test this against ASCII space
            return null;
        }

        // https://stackoverflow.com/questions/18898773/java-escape-json-string
        return new SerializedString(String.format("\\u%04x", i));
    }
}
```

Then:

```java
        JsonFactory f = JsonFactory
                .builder()
                .build()
                .setCharacterEscapes(new EscapeAllNonAsciiCharacters(JsonpCharacterEscapes.instance()));

        this.mapper = JsonMapper
                .builder(f)
                .build();

        this.writer = this.mapper.writer();
```

And later:

```java
requestBody = writer.writeValueAsString(requestData);
```


My `CharacterEscapes` is only called twice (for a much larger JSON object), so I suppose it's not checking escapes for every character, but only some? Is there a way to force it to be called per-character, or is this the wrong abstraction to poke?

**Comment by Ocramius** [comment]
Well, seems like this is a decent workaround, for now:

```java
        this.mapper = new ObjectMapper()
                .enable(JsonGenerator.Feature.ESCAPE_NON_ASCII);
```

Yep, my JSON payload size will suffer, but that's not a concern for the specific application I'm working with.

**Comment by cowtowncoder** [comment]
@Ocramius glad you were able to get things to work well enough.

`CharacterEscapes` is not the easiest (or best documented thing around).
Having said that, there should be some tests in

https://github.com/FasterXML/jackson-core/

(under `src/test/java/` somewhere)

since it's part of streaming component, not databind.

But `JsonWriteFeature.ESCAPE_NON_ASCII` should do the trick for sure.

**Comment by cowtowncoder** [comment]
No plans to implement; closing as part of bigger issue clean up.
